src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,148 @@
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        Node<T> newNode = new Node<>(null, value, null);
+        if (index == 0) {
+            addFirstElement(value, newNode);
+        } else if (index == size) {
+            addTail(newNode);
+        } else {
+            Node<T> nodeByIndex = searchNode(index);
+            Node<T> previousNode = nodeByIndex.previous;
+            newNode = new Node<>(nodeByIndex, value, previousNode);
+            nodeByIndex.previous = newNode;
+            previousNode.next = newNode;
+        }
+        size++;
+    }
+
+    private void addFirstElement(T value, Node<T> newNode) {
+        if (addFirst(newNode)) {
+            Node<T> newHead = new Node<>(head, value, null);
+            head.previous = newHead;
+            head = newHead;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return searchNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> oldNode = searchNode(index);
+        T oldValue = oldNode.value;
+        oldNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> deletedNode = searchNode(index);
+        unlink(deletedNode);
+        return deletedNode.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        while (currentNode != null) {
+            if (currentNode.value == object || (currentNode.value != null
+                    && currentNode.value.equals(object))) {
+                unlink(currentNode);
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException("An incorrect index was entered - " + index);
+        }
+    }
+
+    private Node<T> searchNode(int index) {
+        checkIndex(index);
+        int indexNode = 0;
+        Node<T> currentNode = head;
+        while (indexNode != index) {
+            currentNode = currentNode.next;
+            indexNode++;
+        }
+        return currentNode;
+    }
+
+    private void addTail(Node<T> newNode) {
+        newNode.previous = tail;
+        tail.next = newNode;
+        tail = newNode;
+    }
+
+    private boolean addFirst(Node<T> newNode) {
+        if (size == 0) {
+            head = newNode;
+            tail = newNode;
+            return false;
+        }
+        return true;
+    }
```

Comment:
please refactor in this way:
```suggestion
    private void addFirst(Node<T> newNode) {
            head = newNode;
            tail = newNode;
    }
```